lib/maths/CBoostedTree.cc

```diff
@@ -0,0 +1,1107 @@
+}
+
+double CMse::curvature(double /*prediction*/, double /*actual*/) const {
+    return 2.0;
+}
+
+void CArgMinMse::addImpl(double prediction, double actual) {
+    m_MeanError.add(actual - prediction);
+}
+
+CMse::TArgMinLossUPtr CMse::minimizer() const {
+    return std::make_unique<CArgMinMse>();
+}
+
+double CArgMinMse::value() const {
+    return CBasicStatistics::mean(m_MeanError);
+}
+}
+
+class CBoostedTree::CImpl final {
+public:
+    using TMeanAccumulator = CBasicStatistics::SSampleMean<double>::TAccumulator;
+
+public:
+    CImpl(std::size_t numberThreads, std::size_t dependentVariable, TLossFunctionUPtr loss)
+        : m_NumberThreads{numberThreads},
+          m_DependentVariable{dependentVariable}, m_Loss{std::move(loss)} {}
+
+    //! Set the loss function.
+    //!
+    //! Set the number of folds to use for estimating the generalisation error.
+    void numberFolds(std::size_t k) { m_NumberFolds = k; }
+
+    //! Set the maximum number of trees to use in the forest.
+    void maximumNumberTrees(std::size_t maximumNumberTrees) {
+        m_MaximumNumberTrees = maximumNumberTrees;
+    }
+
+    //! Set the fraction of features which are used in each bag.
+    void featureBagFraction(double featureBagFraction) {
+        m_FeatureBagFraction = featureBagFraction;
+    }
+
+    //! Set the amount by which to shrink the node values for each successive tree
+    //! in the forest.
+    void shrinkageFactor(double shrinkageFactor) {
+        m_ShrinkageFactor = shrinkageFactor;
+    }
+
+    //! Train the model on the values in \p frame.
+    void train(core::CDataFrame& frame, TProgressCallback recordProgress) {
+
+        if (m_Loss == nullptr) {
+            HANDLE_FATAL(<< "Internal error: no loss function defined for regression."
+                         << " Please report this problem.");
+            return;
+        }
+
+        this->initializeMissingFeatureMasks(frame);
+
+        TPackedBitVectorVec trainingRowMasks;
+        TPackedBitVectorVec testingRowMasks;
+        std::tie(trainingRowMasks, testingRowMasks) = this->crossValidationRowMasks();
+
+        // We store the gradient and curvature of the loss function and the predicted
+        // value for the dependent variable of the regression.
+
+        frame.resizeColumns(m_NumberThreads, frame.numberColumns() + 3);
+
+        this->initializeFeatureSampleDistribution(frame);
+
+        this->initializeRegularisation(frame, recordProgress);
+
+        // TODO Hyperparameter optimisation...
+        // Get next parameters to test, compute generalisation error, record if best, repeat.
+
+        LOG_TRACE(<< "Starting training");
+
+        TMeanAccumulator meanLoss;
+        for (std::size_t i = 0; i < m_NumberFolds; ++i) {
+            LOG_TRACE(<< "fold = " << i);
+            TNodeVecVec forest(this->trainForest(frame, trainingRowMasks[i], recordProgress));
+            double loss{this->meanLoss(frame, testingRowMasks[i], forest)};
```

Comment:
I debated this. I think since it is on a hold out set there is no need to additionally penalise model complexity since if it overfit it should generalise badly and the error will be high. It certainly isn't clear cut to me. We can experiment with this down the line though.